### PR TITLE
feat(DLI): add new resource: DLI permission

### LINKED ISSUE #1667: [DLI]Need a new resource to manage Data Permission
### Current Terraform Version

```
...
```

### Proposal

1. grant database or table data usage permission to specified users.
2. grant flink job usage permission to specified users.
3. grant flink package group or package permission to specified users.

### PR DESCRIPTION
**Pull request by hmbig2**
add new resource: DLI permission

**What this PR does / why we need it**:

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1667 

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [X] Tests added/passed.
* [X] Documentation updated.
* [X] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/dli' TESTARGS='-run=TestAccResourceDliAuth_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dli -v -run=TestAccResourceDliAuth_basic -timeout 360m -parallel 4
=== RUN   TestAccResourceDliAuth_basic
=== PAUSE TestAccResourceDliAuth_basic
=== CONT  TestAccResourceDliAuth_basic
--- PASS: TestAccResourceDliAuth_basic (30.11s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dli       30.184s
```
```
make testacc TEST='./huaweicloud/services/acceptance/dli' TESTARGS='-run=TestAccResourceDliAuth'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dli -v -run=TestAccResourceDliAuth -timeout 360m -parallel 4
=== RUN   TestAccResourceDliAuth_basic
=== PAUSE TestAccResourceDliAuth_basic
=== RUN   TestAccResourceDliAuth_flink
=== PAUSE TestAccResourceDliAuth_flink
=== RUN   TestAccResourceDliAuth_table
=== PAUSE TestAccResourceDliAuth_table
=== RUN   TestAccResourceDliAuth_package
=== PAUSE TestAccResourceDliAuth_package
=== RUN   TestAccResourceDliAuth_queue
=== PAUSE TestAccResourceDliAuth_queue
=== CONT  TestAccResourceDliAuth_basic
=== CONT  TestAccResourceDliAuth_package
=== CONT  TestAccResourceDliAuth_table
=== CONT  TestAccResourceDliAuth_flink
--- PASS: TestAccResourceDliAuth_basic (45.97s)
=== CONT  TestAccResourceDliAuth_queue
--- PASS: TestAccResourceDliAuth_table (53.64s)
--- PASS: TestAccResourceDliAuth_package (66.68s)
--- PASS: TestAccResourceDliAuth_flink (94.48s)
--- PASS: TestAccResourceDliAuth_queue (150.23s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dli       196.328s
```

### CODE REVIEW

File: huaweicloud/services/dli/resource_huaweicloud_dli_permission.go [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+	if strings.HasPrefix(obj, "queues.") {
+		queueName := strings.Replace(obj, "queues.", "", 1)
+		opts := auth.GrantQueuePermissionOpts{
+			QueueName: queueName,
+			UserName:  userName,
+			Action:    "update",
+		}
+
+		if privileges, ok := d.GetOk("privileges"); ok {
+			for _, v := range privileges.([]interface{}) {
```
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_permission.go]: it is recommended to use `utils.ExpandToStringList(privileges.([]interface{}))`

File: huaweicloud/services/dli/resource_huaweicloud_dli_permission.go [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+
+	} else {
+		opts := auth.GrantDataPermissionOpts{
+			UserName: userName,
+			Action:   "update",
+		}
+
+		if privileges, ok := d.GetOk("privileges"); ok {
+			var ids []string
+			for _, v := range privileges.([]interface{}) {
```
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_permission.go]: ditto

File: huaweicloud/services/dli/resource_huaweicloud_dli_permission.go [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+			return fmtp.DiagErrorf("Error grant permission in DLI: %s", rst.Message)
+		}
+
+	} else {
+		opts := auth.GrantDataPermissionOpts{
+			UserName: userName,
+			Action:   "update",
+		}
+
+		if privileges, ok := d.GetOk("privileges"); ok {
```
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_permission.go]: `privileges` is a required, it is not necessary for verification

File: huaweicloud/services/dli/resource_huaweicloud_dli_permission.go [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+			}
+			opts.Privileges = append(opts.Privileges, auth.DataPermission{
+				Object:     obj,
+				Privileges: ids,
+			})
+		}
+
+		rst, createErr := auth.GrantDataPermission(client, opts)
+		if createErr != nil {
+			return fmtp.DiagErrorf("Error grant permission in DLI: %s", createErr)
```
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_permission.go:114]: Error granting...

Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_permission.go:114]: ditto

File: huaweicloud/services/dli/resource_huaweicloud_dli_permission.go [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+		if privileges, ok := d.GetOk("privileges"); ok {
+			for _, v := range privileges.([]interface{}) {
+				opts.Privileges = append(opts.Privileges, v.(string))
+			}
+		}
+
+		rst, createErr := auth.GrantQueuePermission(client, opts)
+		if createErr != nil {
+			return fmtp.DiagErrorf("Error grant permission in DLI: %s", createErr)
```
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_permission.go]: Error granting...
also in line: 92, 114, 118

File: huaweicloud/services/dli/resource_huaweicloud_dli_permission.go [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+		}
+
+		if rst != nil && !rst.IsSuccess {
```
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_permission.go:90]: When re-authorizing or modifying resources, is it the same as the first authorization?

File: huaweicloud/services/dli/resource_huaweicloud_dli_permission.go [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+	} else if strings.HasPrefix(obj, "queues") {
+		permission, err = queryQueuePermission(client, obj, userName)
+	} else {
+		return nil, fmtp.Errorf("The object is illegal:object=%s,userName=%s, error=%s", obj, userName, err)
+	}
+
+	return permission, err
+}
+
+func ParseDliErrorToError404(respErr error) error {
```
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_permission.go]: Will this function be used outside this package? If you don't use it, pls change to `parseDliErrorToError404`.

File: huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_permission_test.go [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+				Config: testAccDliAuthResource_flink(name, acceptance.HW_PROJECT_ID, acceptance.HW_REGION_NAME, "START"),
+				Check: resource.ComposeTestCheckFunc(
+					rc.CheckResourceExists(),
+					resource.TestCheckResourceAttr(resourceName, "user_name", name),
+					resource.TestMatchResourceAttr(resourceName, "object", regexp.MustCompile(`jobs\.flink\.\d*`)),
+					resource.TestCheckResourceAttr(resourceName, "privileges.0", "START"),
+					resource.TestCheckResourceAttrSet(resourceName, "is_admin"),
+				),
+			},
+			{
```
Comment (chengxiangdong) [inline huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_permission_test.go:131]: Only keep the import in basic function, and the other does not need to check the import.

File: huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_permission_test.go [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+}
+
+func testAccDliAuthUserConfig(name string, projectId string) string {
+	return fmt.Sprintf(`
+resource "huaweicloud_identity_group" "group_1" {
+  name = "%s"
+}
+
+
+resource "huaweicloud_identity_user" "test" {
```
Comment (chengxiangdong) [inline huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_permission_test.go:351]: one more blank line

Comment (chengxiangdong) [inline huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_permission_test.go:353]: ditto

File: huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_permission_test.go [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+EOF
+}
+
+resource "huaweicloud_identity_role_assignment" "role_assignment_1" {
+  role_id    = huaweicloud_identity_role.role_1.id
+  group_id   = huaweicloud_identity_group.group_1.id
+  project_id = "%s"
+}
+
+
```
Comment (chengxiangdong) [inline huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_permission_test.go]: ditto

File: huaweicloud/services/dli/resource_huaweicloud_dli_permission.go (hunk 7 above)
Comment (hmbig2) [reply huaweicloud/services/dli/resource_huaweicloud_dli_permission.go:340]: yes,used

File: huaweicloud/services/dli/resource_huaweicloud_dli_permission.go (hunk 6 above)
Comment (hmbig2) [reply huaweicloud/services/dli/resource_huaweicloud_dli_permission.go:90]: all authorizate is update permissions

File: docs/resources/dli_permission.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+  + **databases.`database_name`.tables.`table_name`.columns.`column_name`**: the usage permissions of data in the column.
+  + **jobs.flink.`flink_job_id`**: the usage permissions of data in the flink job.
+  + **groups.`package_group_name`**: the usage permissions of data in the package group.
+  + **resources.`package_name`**: the usage permissions of data in the package.
+
+  Changing this parameter will create a new resource.
+
+* `privileges` - (Required, List) Specifies the usage permissions of data.
+  + Permissions on Queue, Database and Table,
+   please see [Permissions Management](https://support.huaweicloud.com/intl/en-us/productdesc-dli/dli_07_0006.html)
```
Comment (Lance52259) [inline docs/resources/dli_permission.md:63]: Maybe `For more details of the permissions on queue, database and table, please see permissions management [document](......).` is better.

File: docs/resources/dli_permission.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+  + Permissions on Flink job.
+    [more](https://support.huaweicloud.com/intl/en-us/usermanual-dli/dli_01_0479.html) :
```
Comment (Lance52259) [inline docs/resources/dli_permission.md]: A single more does not constitute a sentence.

File: docs/resources/dli_permission.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+     * **START**: This permission allows you to start the job.
+     * **STOP**: This permission allows you to stop the job.
+     * **EXPORT**: This permission allows you to export the job.
+     * **GRANT_PRIVILEGE**: This permission allows you to grant job permissions to other users.
+     * **REVOKE_PRIVILEGE**: This permission allows you to revoke the job permissions that other users have but cannot
+      revoke the job creator's permissions.
+     * **SHOW_PRIVILEGES**: This permission allows you to view the job permissions of other users.
+
+  + Permissions on Package Groups and Packages.
+   [more](https://support.huaweicloud.com/intl/en-us/usermanual-dli/dli_01_0477.html) :
```
Comment (Lance52259) [inline docs/resources/dli_permission.md]: ditto.

File: docs/resources/dli_permission.md [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    * **SHOW_PRIVILEGES**: This permission allows you to view the group permissions of other users.
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The resource ID in format of **object/user_name**. It is composed of `object` and `user_name`,
+ separated by a slash.
+
+* `is_admin` - Whether this user is an administrator
```
Comment (Lance52259) [inline docs/resources/dli_permission.md]: Missing a period at the end of sentence.

File: docs/resources/dli_permission.md [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+## Import
+
+The permission can be imported by their `id`, it is composed of `object` and `user_name`, separated by a slash. e.g.:
+
+```
+terraform import huaweicloud_dli_permission.test databases.database_name/user_name
```
Comment (Lance52259) [inline docs/resources/dli_permission.md:104]: Please add a doller ($) in front of sentence.
`$ terraform import ...`

File: huaweicloud/services/dli/resource_huaweicloud_dli_permission.go [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+func checkPrefixMatchDataPermession(obj string) bool {
+	if strings.HasPrefix(obj, "jobs.flink.") {
+		return true
+	}
+	if strings.HasPrefix(obj, "groups.") {
+		return true
+	}
+	if strings.HasPrefix(obj, "resources.") {
+		return true
+	}
```
Comment (Lance52259) [inline huaweicloud/services/dli/resource_huaweicloud_dli_permission.go]: Maybe use `||`?

File: huaweicloud/services/dli/resource_huaweicloud_dli_permission.go [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+
+		if rst != nil && !rst.IsSuccess {
+			return fmtp.DiagErrorf("Error granting permission in DLI: %s", rst.Message)
+		}
+
+	} else {
+		opts := auth.GrantDataPermissionOpts{
+			UserName: userName,
+			Action:   "update",
+		}
```
Comment (Lance52259) [inline huaweicloud/services/dli/resource_huaweicloud_dli_permission.go]: Extractable public code.

File: huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_permission_test.go (hunk 8 above)
Comment (chengxiangdong) [reply huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_permission_test.go:131]: You need to take a look at our specification, it is inconsistent with the specification, although it is not an error.

File: huaweicloud/services/dli/resource_huaweicloud_dli_permission.go (hunk 17 above)
Comment (hmbig2) [reply huaweicloud/services/dli/resource_huaweicloud_dli_permission.go:96]: just parameter ，unnecessary

**Pull request merged**